File: src/main/kotlin/aurocosh/divinefavor/common/entity/minions/minion_interaction/base/MinionInteractionHandler.java
package aurocosh.divinefavor.common.entity.minions.minion_interaction.base;

import aurocosh.divinefavor.common.entity.minions.base.IMinion;
import aurocosh.divinefavor.common.entity.minions.minion_interaction.MinionInteraction;
import net.minecraft.entity.EntityLiving;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumHand;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class MinionInteractionHandler<T extends EntityLiving & IMinion> {
    private final List<MinionInteraction<T>> interactions;
    private final List<MinionInteraction<T>> ownerInteractions;

    public MinionInteractionHandler() {
        this.interactions = new ArrayList<>();
        ownerInteractions = new ArrayList<>();
    }

    public void addInteraction(MinionInteraction<T> interaction) {
        interactions.add(interaction);
    }

    public void addOwnerInteraction(MinionInteraction<T> interaction) {
        ownerInteractions.add(interaction);
    }

    public boolean processInteract(T minion, EntityPlayer player, EnumHand hand) {
        if (player.world.isRemote)
            return false;

        UUID uuid = minion.getMinionData().getOwnerUUID();
        if (player.getGameProfile().getId().equals(uuid))
            for (MinionInteraction<T> interaction : ownerInteractions)
                if (interaction.process(minion, player, hand))
                    return true;

        for (MinionInteraction<T> interaction : interactions)
            if (interaction.process(minion, player, hand))
                return true;
        return false;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/entity/minions/minion_interaction/base/MinionInteractionHandler.kt
package aurocosh.divinefavor.common.entity.minions.minion_interaction.base

import aurocosh.divinefavor.common.entity.minions.base.IMinion
import aurocosh.divinefavor.common.entity.minions.minion_interaction.MinionInteraction
import net.minecraft.entity.EntityLiving
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EnumHand
import java.util.*

class MinionInteractionHandler<T> where T : IMinion, T : EntityLiving {
    private val interactions: MutableList<MinionInteraction<T>>
    private val ownerInteractions: MutableList<MinionInteraction<T>>

    init {
        interactions = ArrayList()
        ownerInteractions = ArrayList()
    }

    fun addInteraction(interaction: MinionInteraction<T>) {
        interactions.add(interaction)
    }

    fun addOwnerInteraction(interaction: MinionInteraction<T>) {
        ownerInteractions.add(interaction)
    }

    fun processInteract(minion: T, player: EntityPlayer, hand: EnumHand): Boolean {
        if (player.world.isRemote)
            return false

        val uuid = minion.minionData.ownerUUID
        if (player.gameProfile.id == uuid)
            for (interaction in ownerInteractions)
                if (interaction.process(minion, player, hand))
                    return true

        for (interaction in interactions)
            if (interaction.process(minion, player, hand))
                return true
        return false
    }
}
